### PR TITLE
CNX-9043 update server info return type and queries to include new migration property

### PR DESCRIPTION
**Pull request by connorivy**
## Description & motivation

Add new account properties to track if a user's account has been upgraded from an FE1 account

## Changes:

Add movedTo and movedFrom props to ServerInfo object
Modify graphQL queries to include the movedFrom and movedTo info

## Checklist:

- [x] My pull request follows the guidelines in the [Contributing guide](https://github.com/specklesystems/speckle-server/blob/main/CONTRIBUTING.md)?
- [x] My pull request does not duplicate any other open [Pull Requests](../../pulls) for the same update/change?
- [x] My commits are related to the pull request and do not amend unrelated code or documentation.
- [x] My code follows a similar style to existing code.
- [ ] I have added appropriate tests.
- [ ] I have updated or added relevant documentation.

### CODE REVIEW

File: Core/Core/Api/GraphQL/GraphQLHttpClientExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+  /// <summary>
+  /// Gets the version of the current server. Useful for guarding against unsupported api calls on newer or older servers.
+  /// </summary>
+  /// <param name="cancellationToken">[Optional] defaults to an empty cancellation token</param>
+  /// <returns><see cref="Version"/> object excluding any strings (eg "2.7.2-alpha.6995" becomes "2.7.2.6995")</returns>
+  /// <exception cref="SpeckleGraphQLException"></exception>
+  public static async Task<System.Version> GetServerVersion(
+    this GraphQLHttpClient client,
+    CancellationToken cancellationToken = default
+  )
```
Comment (AlanRynne) [inline Core/Core/Api/GraphQL/GraphQLHttpClientExtensions.cs:20]: Why an extension method?

File: Core/Core/Credentials/AccountManager.cs [hunk 2]
```diff
@@ -53,8 +54,23 @@ public static async Task<ServerInfo> GetServerInfo(Uri server, CancellationToken
       httpClient
     );
 
+    System.Version version = await gqlClient
+      .GetServerVersion(cancellationToken: cancellationToken)
+      .ConfigureAwait(false);
+    System.Version serverMigrationVersion = new(2, 18, 5);
```
Comment (AlanRynne) [inline Core/Core/Credentials/AccountManager.cs:60]: These are some magical numbers, I wish there was a better way to specify this. If not, we should add a comment saying why 2.18.5 is the threshold (we know now... but whoever comes fix this in the future may not)

File: Core/Core/Credentials/AccountManager.cs [hunk 3]
```diff
@@ -53,8 +54,23 @@ public static async Task<ServerInfo> GetServerInfo(Uri server, CancellationToken
...
+    string queryString;
+    if (version >= serverMigrationVersion)
+    {
+      queryString = "query { serverInfo { name company migration { movedFrom movedTo } } }";
+    }
+    else
+    {
+      queryString = "query { serverInfo { name company } }";
+    }
+
```
Comment (AlanRynne) [inline Core/Core/Credentials/AccountManager.cs:71]: This is for backwards compatibility only, we need to discuss with @BovineOx how long should this stay.

File: Core/Core/Credentials/AccountManager.cs [hunk 4]
```diff
@@ -195,6 +223,25 @@ public static string GetDefaultServerUrl()
...
+    if (!account.serverInfo.url.Contains("https://speckle.xyz"))
+    {
+      throw new SpeckleAccountManagerException($"Can only upgrade accounts on speckle.xyz");
+    }
+
+    account.serverInfo.url = DEFAULT_SERVER_URL;
+    s_accountStorage.UpdateObject(account.id, JsonConvert.SerializeObject(account));
+  }
+
```
Comment (AlanRynne) [inline Core/Core/Credentials/AccountManager.cs:244]: This implementation is **not correct** ⚠️.

We should be using `migration.migratedTo` to determine if an account "can be upgraded" and if so, use said url as the one to swap.

Also, this is out of scope for this ticket, or did I miss something in the past days while i was lost in Digicert? 🙂

File: Core/Core/Helpers/Http.cs [hunk 5]
```diff
@@ -134,7 +134,7 @@ public static async Task<bool> Ping(string hostnameOrAddress)
   /// </summary>
   /// <param name="uri">The URI that should be pinged</param>
   /// <exception cref="HttpRequestException">Request to <paramref name="uri"/> failed</exception>
-  internal static async Task<HttpResponseMessage> HttpPing(Uri uri)
+  public static async Task<HttpResponseMessage> HttpPing(Uri uri)
```
Comment (AlanRynne) [inline Core/Core/Helpers/Http.cs:137]: 🤔 Why are we making this public?

File: Core/Core/Api/GraphQL/Models.cs [hunk 6]
```diff
@@ -396,6 +396,23 @@ public class ServerInfo
...
+  /// One of the base fields for our account upgrade mechanism.
+  /// Shows the url that this account now lives at after upgrading to FE2
+  /// </summary>
+  public Uri movedTo { get; set; }
+
+  /// <summary>
+  /// One of the base fields for our account upgrade mechanism.
+  /// Shows the url that this account used to live at before being upgraded to FE2
+  /// </summary>
+  public Uri movedFrom { get; set; }
```
Comment (AlanRynne) [inline Core/Core/Api/GraphQL/Models.cs:415]: This docs are deceiving. `ServerMigration` class knows nothing of any Account, it's purely server metadata.

`movedFrom` and `movedTo` show the "new URL where this server is now deployed" and "the old RUL where this server used to be deployed" respectively.

File: Core/Core/Api/GraphQL/GraphQLHttpClientExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+      throw new SpeckleGraphQLException<ServerInfoResponse>(
+        $"Query {nameof(GetServerVersion)} failed",
+        request,
+        response
+      );
+    }
+
+    return response.Data.serverInfo.version.Contains("dev")
+      ? new System.Version(999, 999, 999)
+      : new System.Version(Regex.Replace(response.Data.serverInfo.version, "[-a-zA-Z]+", ""));
```
Comment (AlanRynne) [inline Core/Core/Api/GraphQL/GraphQLHttpClientExtensions.cs:45]: I think you need to run me through this 🤔 If we're checking if it contains `dev`, why are we then removing all A-Z characters with regex.

Do we know if `dev` is the only possible edge-case?

**Review by AlanRynne — CHANGES_REQUESTED** [review]
I've got many questions hehe 😅🙌🏼

File: Core/Core/Credentials/AccountManager.cs (hunk 4 above)
Comment (JR-Morgan) [reply Core/Core/Credentials/AccountManager.cs:244]: I think this might just be matteo original implementation, Connors pr is still draft, so I imagine there's more to come.

File: Core/Core/Api/GraphQL/GraphQLHttpClientExtensions.cs (hunk 1 above)
Comment (connorivy) [reply Core/Core/Api/GraphQL/GraphQLHttpClientExtensions.cs:20]: I like extensions methods. They are reusable, discoverable with intellisense, and organized so we'll never end up with massive utils files. I didn't want to add it as a method on line 850 of AccountManager so this is what I came up with, but I'm happy to change it if you have something else in mind.

File: Core/Core/Credentials/AccountManager.cs (hunk 3 above)
Comment (connorivy) [reply Core/Core/Credentials/AccountManager.cs:71]: this is indeed only for backwards compatibility

File: Core/Core/Api/GraphQL/GraphQLHttpClientExtensions.cs (hunk 7 above)
Comment (connorivy) [reply Core/Core/Api/GraphQL/GraphQLHttpClientExtensions.cs:45]: This logic was taken from Client.ServerOperations.cs

Unfortunately, I could just use that implementation because it only applies to a Speckle.Core.Api.Client which is coupled to our Account class. Where this method is being used currently is a static context that is ignorant of any accounts so we can't use the existing logic.

File: Core/Core/Api/GraphQL/Models.cs (hunk 6 above)
Comment (connorivy) [reply Core/Core/Api/GraphQL/Models.cs:415]: You're right this is misleading

File: Core/Core/Api/GraphQL/GraphQLHttpClientExtensions.cs (hunk 1 above)
Comment (AlanRynne) [reply Core/Core/Api/GraphQL/GraphQLHttpClientExtensions.cs:20]: yeah i do agree 850 lines seems like enough. But extension methods were really created to extend a class you don't have access to (i.e from a dependency), this seems a bit of a misuse, but I accept your reasoning :)

File: Core/Core/Api/GraphQL/GraphQLHttpClientExtensions.cs (hunk 7 above)
Comment (AlanRynne) [reply Core/Core/Api/GraphQL/GraphQLHttpClientExtensions.cs:45]: So you're saying we now have this weird logic in 2 places? 😅

I think we should check with someone from the server team what are the edge-cases expected in the serverInfo.version return value.

File: Core/Core/Credentials/AccountManager.cs (hunk 3 above)
Comment (AlanRynne) [reply Core/Core/Credentials/AccountManager.cs:74]: I'm opening a ticket for this and letting Ian know

File: Core/Core/Api/GraphQL/GraphQLHttpClientExtensions.cs (hunk 7 above)
Comment (JR-Morgan) [reply Core/Core/Api/GraphQL/GraphQLHttpClientExtensions.cs:45]: Can we have a quick chat on this today? It would be nice if we could avoid duplicating this logic

Comment (JR-Morgan) [reply Core/Core/Api/GraphQL/GraphQLHttpClientExtensions.cs:45]: From our discussion with @didimitrie and @gjedlicska :
 - The other server version check can be killed, was only for 2.6.0 backwards compact
 - The `.Contains("Dev")` can be replaced by a `== "Dev"`

There's a couple other topics I want chat about next week in Prague r.e. our `Client` class. But for now, just the above ^ affects this PR.

**Review by AlanRynne — APPROVED** [review]
